x/merkledb/trieview.go

```diff
@@ -183,61 +178,50 @@ func newTrieViewWithChanges(
 		return nil, ErrNoValidRoot
 	}
 
-	return &trieView{
+	newView := &trieView{
 		root:       passedRootChange.after,
 		db:         db,
 		parentTrie: parentTrie,
 		changes:    changes,
-	}, nil
+	}
+	// the changes have already been calculated, don't need to do it again later
+	newView.calculateNodesOnce.Do(func() {})
+	return newView, nil
 }
 
 // Recalculates the node IDs for all changed nodes in the trie.
-// Assumes [t.lock] is held.
 func (t *trieView) calculateNodeIDs(ctx context.Context) error {
-	switch {
-	case t.isInvalid():
-		return ErrInvalid
-	case !t.needsRecalculation:
-		return nil
-	case t.committed:
-		// Note that this should never happen. If a view is committed, it should
-		// never be edited, so [t.needsRecalculation] should always be false.
-		return ErrCommitted
-	}
-
-	// We wait to create the span until after checking that we need to actually
-	// calculateNodeIDs to make traces more useful (otherwise there may be a span
-	// per key modified even though IDs are not re-calculated).
-	ctx, span := t.db.tracer.Start(ctx, "MerkleDB.trieview.calculateNodeIDs")
-	defer span.End()
-
-	// ensure that the view under this one is up-to-date before potentially pulling in nodes from it
-	// getting the Merkle root forces any unupdated nodes to recalculate their ids
-	if _, err := t.getParentTrie().GetMerkleRoot(ctx); err != nil {
-		return err
-	}
-
-	_, helperSpan := t.db.tracer.Start(ctx, "MerkleDB.trieview.calculateNodeIDsHelper")
-	defer helperSpan.End()
-
-	// [eg] limits the number of goroutines we start.
-	var eg errgroup.Group
-	eg.SetLimit(numCPU)
-	if err := t.calculateNodeIDsHelper(ctx, t.root, &eg); err != nil {
-		return err
-	}
-	if err := eg.Wait(); err != nil {
-		return err
-	}
-	t.needsRecalculation = false
-	t.changes.rootID = t.root.id
+	var err error
+	t.calculateNodesOnce.Do(func() {
+		if t.isInvalid() {
+			err = ErrInvalid
+			return
+		}
 
-	// ensure no ancestor changes occurred during execution
-	if t.isInvalid() {
-		return ErrInvalid
-	}
+		// We wait to create the span until after checking that we need to actually
+		// calculateNodeIDs to make traces more useful (otherwise there may be a span
+		// per key modified even though IDs are not re-calculated).
+		ctx, span := t.db.tracer.Start(ctx, "MerkleDB.trieview.calculateNodeIDs")
+		defer span.End()
+
+		// [eg] limits the number of goroutines we start.
+		var eg errgroup.Group
+		eg.SetLimit(numCPU)
+		if err = t.calculateNodeIDsHelper(ctx, t.root, &eg); err != nil {
+			return
+		}
+		if err = eg.Wait(); err != nil {
+			return
+		}
+		t.changes.rootID = t.root.id
```

Comment:
Is it a problem that we're setting `t.changes` here while someone could simultaneously accessing it in, for example, `getNodeWithID`?

Reply:
Shouldn't be an issue, the only way to trigger a call to getNodeWithID is during inserts into the trie, which happens within this function, or from a child view, which cannot happen before this function completes.

Reply:
>  the only way to trigger a call to getNodeWithID is during inserts into the trie
I don't think this is true. What about `getProof`?

Also in `getValue` we use `t.changes` in `getValue`

Reply:
t.changes.values doesn't change once the view is created so it never can be written to during a read and getProof also calls calculateNodeIDs before running so all writes will finish before being read.

Reply:
> t.changes.values doesn't change once the view is created

it also changes in `commitChanges`, no?